### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [2]:
tot_players= purchase_data['SN'].value_counts()
tot_players= len(tot_players)
tot_players

576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
#unique items, average price
tottyp_uniq= purchase_data['Item ID'].unique()
tottyp_uniq=len(tottyp_uniq)
#tottyp_uniq
#average price
ave_pri= purchase_data['Price'].mean()
#ave_pri
#total purchase
tot_pur= purchase_data['Purchase ID'].count()
#tot_pur
#total Revenue
tot_rev= purchase_data['Price'].sum()
#tot_rev

#create a dataframe

totpur_anly= pd.DataFrame({'Total Players':[tot_players],
                            'Total Uniques':[tottyp_uniq],
                            'Average Price':["${:,.2f}".format(ave_pri)],
                            'Total Purchases':[tot_pur],
                            'Total Revenue':["${:,.2f}".format(tot_rev)]})
totpur_anly=totpur_anly.set_index('Total Players')
totpur_anly

Total Uniques Average Price  Total Purchases Total Revenue
Total Players                                                            
576                      179         $3.05              780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:
#A data frame with total player
#tot_pla = purchase_data['SN'].value_counts()
#gend= {'Player':tot_pla, 'Gender':purchase_data['Gender']}
gend_df= pd.DataFrame({'Player':purchase_data['SN'], 
                     'Gender':purchase_data['Gender']})


gend_df=gend_df.drop_duplicates()

tot_gend= len(gend_df['Gender'])

tot_male=0
tot_fem=0
tot_oth=0

for gend in gend_df['Gender']:
    if gend == 'Male':
        tot_male =tot_male+1 
    elif gend == 'Female':
        tot_fem =tot_fem+1 
    elif gend == 'Other / Non-Disclosed':
        tot_oth =tot_oth+1 


        
gend_sumdf = pd.DataFrame({'Gender': ["Male", "Female", "Other / Non-Disclosed"],
                          'Total Counts':[tot_male, tot_fem, tot_oth],
                          'percentage':['{:,.2%}'.format(tot_male/tot_gend), 
                                        '{:,.2%}'.format(tot_fem/tot_gend), 
                                        '{:,.2%}'.format(tot_oth/tot_gend)]})
gend_sumdf = gend_sumdf.set_index('Gender')
gend_sumdf
# tot_oth
# len(tot_fem)
# tot_oth

Total Counts percentage
Gender                                        
Male                            484     84.03%
Female                           81     14.06%
Other / Non-Disclosed            11      1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
#lets write a series for purchase count
#Purchase Count 	Average Purchase Price 	Total Purchase Value 	Avg Total Purchase per Person
tot_malepur=0
tot_fempur=0
tot_othpur=0



for gend in purchase_data['Gender']:
    if gend == 'Male':
        tot_malepur =tot_malepur+1 
    elif gend == 'Female':
        tot_fempur= tot_fempur+1
    elif gend == 'Other / Non-Disclosed':
        tot_othpur =tot_othpur+1
purchase_count=[tot_malepur, tot_fempur,tot_othpur]
#Purchase Value and Average Purchase Price
male_pri= purchase_data.loc[purchase_data["Gender"] == "Male", ["Price"]]
tot_maleval= round(sum(male_pri['Price']), 2)
fem_pri= purchase_data.loc[purchase_data["Gender"] == "Female", ["Price"]]
tot_femval= round(sum(fem_pri['Price']), 2)
oth_pri= purchase_data.loc[purchase_data["Gender"] == "Other / Non-Disclosed", ["Price"]]
tot_othval =round(sum(oth_pri['Price']), 2)

purchase_value=["${:,.2f}".format(tot_maleval),"${:,.2f}".format(tot_femval),"${:,.2f}".format(tot_othval)]
ave_pur_pri =["${:,.2f}".format(tot_maleval/tot_malepur),"${:,.2f}".format(tot_femval/tot_fempur),"${:,.2f}".format(tot_othval/tot_othpur)]
ave_pur_pers=["${:,.2f}".format(tot_maleval/tot_male),"${:,.2f}".format(tot_femval/tot_fem),"${:,.2f}".format(tot_othval/tot_oth)]
#New Dataframe
gend_anly = pd.DataFrame({'Gender': ["Male", "Female", "Other / Non-Disclosed"],
                          'Purchase Count':purchase_count,
                          'Average Purchase Price':ave_pur_pri,
                          'Total Purchase Value':purchase_value,
                          'Avg Total Purchase per Person':ave_pur_pers})
gend_anly.set_index('Gender')

Purchase Count Average Purchase Price  \
Gender                                                         
Male                              652                  $3.02   
Female                            113                  $3.20   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Male                             $1,967.64                         $4.07  
Female                             $361.94                         $4.47  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [6]:
age_bins=[0, 9,14,19,24,29,34,39,max(purchase_data["Age"])]
age_labels=['<10', '10-14','15-19','20-24','25-29','30-34','35-39','40+']
age_series = pd.cut(purchase_data["Age"], age_bins, labels=age_labels, include_lowest=True)
# age_df=pd.DataFrame({'Age Group':age_labels})
purchase_data["Age Group"]=age_series
ageg_df=pd.DataFrame({'Age Group' : purchase_data["Age Group"],
                          'Total Count' : purchase_data["SN"]})
#del(uni_purdata['Purchase ID']#,'Gender','Item ID','Item Name','Price'])
ageg_df=ageg_df.set_index('Age Group')
ageg_df=ageg_df.drop_duplicates()
ageg_df=ageg_df.groupby("Age Group").count()
ageg_df['Percentages']= ageg_df['Total Count']/ageg_df['Total Count'].sum()
ageg_df['Percentages']=['{:,.2%}'.format(agep) for agep in ageg_df['Percentages']]
ageg_df

Total Count Percentages
Age Group                         
<10                 17       2.95%
10-14               22       3.82%
15-19              107      18.58%
20-24              258      44.79%
25-29               77      13.37%
30-34               52       9.03%
35-39               31       5.38%
40+                 12       2.08%

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
#Purchase Count 	Average Purchase Price 	Total Purchase Value 	Avg Total Purchase per Person
#ageg_anly=purchase_data.groupby('Age Group')
Pur_countbyAge=purchase_data.groupby('Age Group').count()['SN']
#purchase per person is from another dataset with the same indent
PurPers_countbyAge=ageg_df['Total Count']
Ave_pricebyAge=purchase_data.groupby('Age Group').mean()['Price']
Ave_pricebyAge=["${:,.2f}".format(apba) for apba in Ave_pricebyAge]
Tot_pricebyAge=purchase_data.groupby('Age Group').sum()['Price']
Tot_byAgeformat=["${:,.2f}".format(tpba) for tpba in Tot_pricebyAge]
APP_age= Tot_pricebyAge/PurPers_countbyAge
APP_ageformat=["${:,.2f}".format(appa) for appa in APP_age]
ageg_anly = pd.DataFrame({'Purchase Count':Pur_countbyAge,
                          'Average Purchase Price':Ave_pricebyAge,
                          'Total Purchase Value':Tot_byAgeformat,
                          'Average Total Purchase':APP_ageformat})

ageg_anly

Purchase Count Average Purchase Price Total Purchase Value  \
Age Group                                                               
<10                    23                  $3.35               $77.13   
10-14                  28                  $2.96               $82.78   
15-19                 136                  $3.04              $412.89   
20-24                 365                  $3.05            $1,114.06   
25-29                 101                  $2.90              $293.00   
30-34                  73                  $2.93              $214.00   
35-39                  41                  $3.60              $147.67   
40+                    13                  $2.94               $38.24   

          Average Total Purchase  
Age Group                         
<10                        $4.54  
10-14                      $3.76  
15-19                      $3.86  
20-24                      $4.32  
25-29                      $3.81  
30-34                      $4.12  
35-39                      $4.76  
40+                        $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [8]:
#use a sort # age is a placeholder for any kind of count
spsn_groupby=purchase_data.groupby('SN').count()
spsn_groupby=spsn_groupby.reset_index()
spsn_groupby=spsn_groupby.sort_values('Age', ascending=False)
del(spsn_groupby['Gender'],spsn_groupby['Item ID'],spsn_groupby['Item Name'],spsn_groupby['Age Group'],spsn_groupby['Price'])

price_sort=purchase_data.groupby('SN').sum()
price_sort=price_sort.reset_index()
price_sort=price_sort.sort_values('Price', ascending=False)

del(price_sort['Age'],price_sort['Item ID'])

merge_df=pd.merge(spsn_groupby, price_sort, on="SN")

apts = merge_df['Price']/merge_df['Age']


sp_df=pd.DataFrame({'SN':merge_df['SN'],
                     'Total Purchases': merge_df['Age'],
                     'Average Per Purchase':apts.map("${:,.2f}".format),
                     'Total Paid':merge_df['Price']})

sp_df=sp_df.sort_values('Total Paid', ascending=False)

sp_df['Total Paid']=sp_df['Total Paid'].map("${:,.2f}".format)
sp_df


#Average Purchase Price 	Total Purchase Value
#Maybe we used Merge?
# merge_df=pd.merge(purchase_data, sp_df, on="SN")
# del(merge_df['Age'],merge_df['Gender'],merge_df['Item ID'],merge_df['Item Name'],merge_df['Age Group'])
# sp_sort=merge_df.groupby('SN').sum()


SN  Total Purchases Average Per Purchase Total Paid
0        Lisosia93                5                $3.79     $18.96
2      Idastidru52                4                $3.86     $15.45
26      Chamjask73                3                $4.61     $13.83
1           Iral74                4                $3.40     $13.62
32     Iskadarya95                3                $4.37     $13.10
..             ...              ...                  ...        ...
524  Frichjaskan98                1                $1.02      $1.02
487      Isurria36                1                $1.02      $1.02
341     Chanirra79                1                $1.01      $1.01
400        Aidai61                1                $1.01      $1.01
444          Alo38                1                $1.00      $1.00

[576 rows x 4 columns]

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
item_df= pd.DataFrame({'Item ID':purchase_data['Item ID'],
                       'Item Name':purchase_data['Item Name'],
                       'Price':purchase_data['Price']})
item_gb=item_df.groupby(['Item ID', 'Item Name'])
item_gb=item_gb.count().reset_index()
item_gb=item_gb.rename(columns={'Price':'Purchase Counts'})
item_gb=item_gb.sort_values('Purchase Counts', ascending=False)
item_gb

item_df=pd.merge(item_gb, item_df, on=['Item Name','Item ID'], how='left')
item_df=item_df.groupby('Item Name', as_index=False).mean()
item_df=item_df.sort_values('Purchase Counts', ascending=False)


tot_revitem= item_df['Price'] * item_df['Purchase Counts']
item_df['Total Purchase Value(USD)']=tot_revitem
item_df=item_df.rename(columns={'Price':'Price(USD)'})
#item_df['Price(USD)']=round(item_df['Price(USD)'], 2)
item_df['Price(USD)']=item_df['Price(USD)'].map("${:,.2f}".format)
item_df['Total Purchase Value(USD)']=item_df['Total Purchase Value(USD)'].map("${:,.2f}".format)
item_df
# item_df['Total Purchase Value']=item_df['Total Purchase Value'].map("${:,.2f}".format)
# item_df['Price']=item_df['Price'].map("${:,.2f}".format)

Item Name  Item ID  Purchase Counts  \
56                                   Final Critic       92               13   
93   Oathbreaker, Last Hope of the Breaking Storm      178               12   
98                                     Persuasion      132                9   
92                                        Nirvana       82                9   
51      Extraction, Quickblade Of Trembling Hands      108                9   
..                                            ...      ...              ...   
114               Riddle, Tribute of Ended Dreams       27                1   
12                                       Betrayer       90                1   
145                               The Decapitator       42                1   
67                             Gladiator's Glaive      104                1   
66               Ghost Reaver, Longsword of Magic      118                1   

    Price(USD) Total Purchase Value(USD)  
56       $4.61                    $59.99  
93       $4.23                    $50.76  
98       $3.22                    $28.99  
92       $4.90                    $44.10  
51       $3.53                    $31.77  
..         ...                       ...  
114      $3.30                     $3.30  
12       $2.94                     $2.94  
145      $1.75                     $1.75  
67       $1.93                     $1.93  
66       $2.17                     $2.17  

[179 rows x 5 columns]

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [10]:
# #DAMN STRINGS HOW TO SORT CURRECNY/use slack overflow solution,
item_df[['Total Purchase Value(USD)']] = item_df[['Total Purchase Value(USD)']].replace('[\$,]','',regex=True).astype(float)
item_df=item_df.sort_values('Total Purchase Value(USD)', ascending=False)
item_df['Total Purchase Value(USD)']=item_df['Total Purchase Value(USD)'].map("${:,.2f}".format)
item_df

Item Name  Item ID  Purchase Counts  \
56                                   Final Critic       92               13   
93   Oathbreaker, Last Hope of the Breaking Storm      178               12   
92                                        Nirvana       82                9   
55                           Fiery Glass Crusader      145                9   
125                                Singed Scalpel      103                8   
..                                            ...      ...              ...   
57               Flux, Destroyer of Due Diligence       28                2   
49                       Exiled Mithril Longsword      126                1   
170                    Whistling Mithril Warblade      125                2   
67                             Gladiator's Glaive      104                1   
145                               The Decapitator       42                1   

    Price(USD) Total Purchase Value(USD)  
56       $4.61                    $59.99  
93       $4.23                    $50.76  
92       $4.90                    $44.10  
55       $4.58                    $41.22  
125      $4.35                    $34.80  
..         ...                       ...  
57       $1.06                     $2.12  
49       $2.00                     $2.00  
170      $1.00                     $2.00  
67       $1.93                     $1.93  
145      $1.75                     $1.75  

[179 rows x 5 columns]

THREE OBSERVATIONS:

1. Most purchasers of mobile games items are Males(%84). 
2. Most purchasers of both Genders are betweem aged 15-35, Purchasing(approximate 68%), the peak is at their 20s (approximate %45).
4. In first glance, higher price items are bought more than lower price ones, quality of the uniques is more important than price. 
5. 576 peoples bought a total times of 780 	items, only 1 player bought more than 4 items. It depend on how 'recent' the games been launched and the percentages of the playerbase, whether or not the 179 designs for the uniques can be said to be worth the cost and efforts of development.